client/state-db/src/noncanonical.rs

```diff
@@ -432,19 +461,49 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 		})
 	}
 
+	/// Revert a single block. Returns commit set that deletes the journal or `None` if not possible.
+	pub fn remove(&mut self, hash: &BlockHash) -> Option<CommitSet<Key>> {
+		let mut commit = CommitSet::default();
+		let level_count = self.levels.len();
+		for (level_index, level) in self.levels.iter_mut().enumerate().rev() {
+			let index = match level.blocks.iter().position(|overlay| &overlay.hash == hash) {
+				Some(index) => index,
+				None => continue,
+			};
+			// Check that it does not have any children
+			if (level_index != level_count - 1) && self.parents.values().any(|h| h == hash) {
+				log::debug!(target: "state-db", "Trying to remove block {:?} with children", hash);
+				return None;
+			}
+			let overlay = level.remove(index);
+			commit.meta.deleted.push(overlay.journal_key);
+			self.parents.remove(&overlay.hash);
+			discard_values(&mut self.values, overlay.inserted);
```

Comment:
```suggestion
			discard_values(&mut self.values, overlay.inserted);
			break;
```

We should be able to break here? As the block should only exists once?